### DISCUSSION TITLE
Verifiable Certification Schema Decision

### DISCUSSION THREAD

**Discussion by oriolcanades** [DOME-Marketplace/architecture#3] [ADR]
# Verifiable Certification Schema Decision

## Status

The current status of this decision is **PROPOSED**.

## Context

To guarantee that the services offered through the DOME Platform comply with European Union regulations, we need to provide a way to verify the certifications of these services. The Verifiable Credential Schema must be a standard that enables the creation, issuance, and verification of certification credentials in a secure and privacy-preserving way.

In the initial meetings with Tecnalia, Dekra, Ficodes, and IN2, we decided that the first step toward achieving this goal is to create a schema for Verifiable Certifications. This schema will define the structure of the certification credentials that will be issued and verified by services offered through the DOME Marketplace.

### Proposed Schema

The proposed schema for the Verifiable Certification is the following:

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "title": "Verifiable Certification",
  "description": "Defines the structure for a Verifiable Certification Credential issued by DEKRA",
  "type": "object",
  "properties": {
    "@context": {
      "description": "Semantic context for the issued certification",
      "type": "array",
      "items": {
        "type": "string",
        "format": "uri"
      }
    },
    "id": {
      "description": "Globally unique identifier for the credential",
      "type": "string",
      "format": "uri"
    },
    "type": {
      "description": "Full type chain, used to identify the credential base types",
      "type": "array",
      "items": {
        "type": "string"
      }
    },
    "issuer": {
      "description": "Details about the entity issuing the credential",
      "type": "object",
      "properties": {
        "commonName": {
          "description": "Common name of the issuer",
          "type": "string"
        },
        "country": {
          "description": "Country of the issuer",
          "type": "string"
        },
        "id": {
          "description": "DID of the credential issuer",
          "type": "string",
          "format": "uri"
        },
        "organization": {
          "description": "Organization of the issuer",
          "type": "string"
        }
      }
    },
    "issuanceDate": {
      "description": "Date and time when the credential was issued",
      "type": "string",
      "format": "date-time"
    },
    "validFrom": {
      "description": "Date and time when the credential becomes valid",
      "type": "string",
      "format": "date-time"
    },
    "expirationDate": {
      "description": "Date and time when the credential expires",
      "type": "string",
      "format": "date-time"
    },
    "credentialSubject": {
      "description": "Details about the subject of the certification",
      "type": "object",
      "properties": {
        "company": {
          "description": "Details of the company receiving the certification",
          "type": "object",
          "properties": {
            "address": {
              "description": "Address of the company",
              "type": "string"
            },
            "commonName": {
              "description": "Common name of the company",
              "type": "string"
            },
            "country": {
              "description": "Country of the company",
              "type": "string"
            },
            "email": {
              "description": "Email address of the company",
              "type": "string",
              "format": "email"
            },
            "id": {
              "description": "DID of the company",
              "type": "string",
              "format": "uri"
            },
            "organization": {
              "description": "Organization name of the company",
              "type": "string"
            }
          }
        },
        "compliance": {
          "description": "List of compliance certifications held by the company",
          "type": "array",
          "items": {
            "type": "object",
            "properties": {
              "id": {
                "description": "Unique identifier for the compliance certification",
                "type": "string"
              },
              "scope": {
                "description": "Scope of the compliance certification",
                "type": "string"
              },
              "standard": {
                "description": "Standard followed for the compliance certification",
                "type": "string"
              }
            }
          }
        },
        "product": {
          "description": "Details about the certified product",
          "type": "object",
          "properties": {
            "productId": {
              "description": "Unique identifier for the product",
              "type": "string"
            },
            "productName": {
              "description": "Name of the certified product",
              "type": "string"
            },
            "productVersion": {
              "description": "Version of the certified product",
              "type": "string"
            }
          }
        }
      }
    },
    "signer": {
      "description": "Details about the signer of the certification",
      "type": "object",
      "properties": {
        "commonName": {
          "description": "Common name of the signer",
          "type": "string"
        },
        "country": {
          "description": "Country of the signer",
          "type": "string"
        },
        "emailAddress": {
          "description": "Email address of the signer",
          "type": "string",
          "format": "email"
        },
        "organization": {
          "description": "Organization of the signer",
          "type": "string"
        },
        "organizationIdentifier": {
          "description": "Identifier of the organization",
          "type": "string"
        },
        "serialNumber": {
          "description": "Serial number of the signer",
          "type": "string"
        }
      }
    }
  },
  "required": [
    "@context",
    "id",
    "type",
    "issuer",
    "issuanceDate",
    "validFrom",
    "expirationDate",
    "credentialSubject"
  ]
}
```

### Attributes of the schema

| **Name**                                   | **Required** | **Type**         | **Description**                                       |
|--------------------------------------------|--------------|------------------|-------------------------------------------------------|
| `@context`                                 | Yes          | Array of Strings | Semantic context for the issued certification         |
| `id`                                       | Yes          | String           | Globally unique identifier for the credential         |
| `type`                                     | Yes          | Array of Strings | Full type chain to identify the credential base types |
| `issuer`                                   | Yes          | Object           | Details about the entity issuing the credential       |
| `issuer.commonName`                        | No           | String           | Common name of the issuer                             |
| `issuer.country`                           | No           | String           | Country of the issuer                                 |
| `issuer.id`                                | No           | String           | DID of the credential issuer                          |
| `issuer.organization`                      | No           | String           | Organization of the issuer                            |
| `issuanceDate`                             | Yes          | String           | Date and time when the credential was issued          |
| `validFrom`                                | Yes          | String           | Date and time when the credential becomes valid       |
| `expirationDate`                           | Yes          | String           | Date and time when the credential expires             |
| `credentialSubject`                        | Yes          | Object           | Details about the subject of the certification        |
| `credentialSubject.company.address`        | No           | String           | Address of the company                                |
| `credentialSubject.company.commonName`     | No           | String           | Common name of the company                            |
| `credentialSubject.company.country`        | No           | String           | Country of the company                                |
| `credentialSubject.company.email`          | No           | String           | Email address of the company                          |
| `credentialSubject.company.id`             | No           | String           | DID of the company                                    |
| `credentialSubject.company.organization`   | No           | String           | Organization name of the company                      |
| `credentialSubject.compliance`             | No           | Array of Objects | List of compliance certifications held                |
| `credentialSubject.compliance.id`          | No           | String           | Unique identifier for the compliance cert.            |
| `credentialSubject.compliance.scope`       | No           | String           | Scope of the compliance certification                 |
| `credentialSubject.compliance.standard`    | No           | String           | Standard followed for the certification               |
| `credentialSubject.product`                | No           | Object           | Details about the certified product                   |
| `credentialSubject.product.productId`      | No           | String           | Unique identifier for the product                     |
| `credentialSubject.product.productName`    | No           | String           | Name of the certified product                         |
| `credentialSubject.product.productVersion` | No           | String           | Version of the certified product                      |
| `signer`                                   | No           | Object           | Details about the signer of the certification         |
| `signer.commonName`                        | No           | String           | Common name of the signer                             |
| `signer.country`                           | No           | String           | Country of the signer                                 |
| `signer.emailAddress`                      | No           | String           | Email address of the signer                           |
| `signer.organization`                      | No           | String           | Organization of the signer                            |
| `signer.organizationIdentifier`            | No           | String           | Identifier of the organization                        |
| `signer.serialNumber`                      | No           | String           | Serial number of the signer                           |

### Examples

These are several Verifiable Certification examples:

```json
{
  "@context": [
    "https://www.w3.org/ns/credentials/v2",
    "https://dome-marketplace.eu/2022/credentials/learcredential/v1"
  ],
  "id": "72759bb1-5918-4f3b-8520-33359315286d",
  "type": [
    "VerifiableCertification",
    "VerifiableCredential"
  ],
  "issuer": {
    "commonName": "DEKRA",
    "country": "ES",
    "id": "did:elsi:VATES-A29507456",
    "organization": "DEKRA Testing and Certification, S.A.U."
  },
  "credentialSubject": {
    "company": {
      "address": "Fabryczna 5, 00-446 Warszawa",
      "commonName": "CloudFerro",
      "country": "POL",
      "email": "info@cloudferro.com",
      "id": "did:elsi:VATPL-7010468205",
      "organization": "CloudFerro S.A."
    },
    "compliance": [
      {
        "id": "00f75132-ae99-4758-94c5-5373887b35a0",
        "scope": "Information security management systems",
        "standard": "ISO/IEC 27001"
      },
      {
        "id": "e2ea0223-a393-4bed-8316-0a1ae8839086",
        "scope": "Quality management systems",
        "standard": "ISO 9001"
      },
      {
        "id": "3b8e3157-88a1-4cc6-9230-3471c5fa86ed",
        "scope": "Cloud Computing Compliance Criteria Catalogue",
        "standard": "BSI C5"
      }
    ],
    "product": {
      "productId": "urn:ngsi-ld:product-specification:2a32ab98-4104-49bf-97b7-d0d0e1477b45",
      "productName": "CloudFerro Cloud",
      "productVersion": "1.0"
    }
  },
  "issuanceDate": "2024-09-18T17:11:00Z",
  "validFrom": "2024-09-18T17:11:00Z",
  "expirationDate": "2024-10-02T23:59:59Z",
  "signer": {
    "commonName": "DOME DIGITAL IDENTITY",
    "country": "ES",
    "emailAddress": "domesupport@in2.es",
    "organization": "IN2",
    "organizationIdentifier": "VATEU-B99999999",
    "serialNumber": "IDCES-56565656P"
  }
}
```

```json
{
    "@context": [
      "https://www.w3.org/ns/credentials/v2",
      "https://dome-marketplace.eu/2022/credentials/learcredential/v1"
    ],
    "id": "c27143ce-c1df-4824-83bf-b9fc65f3aa38",
    "type": [
      "VerifiableCertification",
      "VerifiableCredential"
    ],
    "issuer": {
      "commonName": "DEKRA",
      "country": "ES",
      "id": "did:elsi:VATES-A29507456",
      "organization": "DEKRA Testing and Certification, S.A.U."
    },
    "credentialSubject": {
      "company": {
        "address": "Viale Carlo Viola, 76 - 11026 Pont-Saint-Martin (AO)",
        "commonName": "ENG",
        "country": "IT",
        "email": "info.dhub@eng.it",
        "id": "did:elsi:VATIT-05724831002",
        "organization": "Engineering DHUB S.p.A."
      },
      "compliance": [
        {
          "id": "d4f4f56f-3722-44f7-a019-1ff4f3f0e397",
          "scope": "Quality management systems",
          "standard": "ISO 9001"
        },
        {
          "id": "6711ce68-0e1b-4f4a-8093-df53ecf3cd4d",
          "scope": "Quality of IT services management",
          "standard": "ISO/IEC 20000-1"
        },
        {
          "id": "ce0ab197-786d-4630-a3bf-c744962ba2d1",
          "scope": "Information security management systems",
          "standard": "ISO/IEC 27001"
        },
        {
          "id": "b32063d0-5310-4597-a3e1-0db5a289d622",
          "scope": "Anti-bribery management systems",
          "standard": "ISO 37001"
        },
        {
          "id": "b1ba7e6f-83b8-498d-b616-cea07b2a8d9f",
          "scope": "Business Continuity Management System",
          "standard": "ISO 22301"
        }
      ],
      "product": {
        "productId": "urn:ngsi-ld:product-specification:59f5e14d-a3ce-4faa-9cfb-d3a03e1b212f",
        "productName": "Availability Protectt",
        "productVersion": "0.1"
      }
    },
    "issuanceDate": "2024-09-18T17:11:00Z",
    "validFrom": "2024-09-18T17:11:00Z",
    "expirationDate": "2025-07-17T23:59:59Z",
    "signer": {
      "commonName": "DOME DIGITAL IDENTITY",
      "country": "ES",
      "emailAddress": "domesupport@in2.es",
      "organization": "IN2",
      "organizationIdentifier": "VATEU-B99999999",
      "serialNumber": "IDCES-56565656P"
    }
}
```

Now, its time to decide if we will use this schema for the DOME Marketplace or we need tu update it in some way.

## Decision

> TODO:

## Consequences

We need to keep in mind that, as of the time of publishing this discussion, 23 Verifiable Certifications have been issued and published in the DOME Marketplace. If we decide to change the schema, all existing certifications will need to be updated to ensure compliance with the new schema.

**Comment by oriolcanades** [comment]
Added more VerifiebleCertification examples
[verifiableCertification1.json](https://github.com/user-attachments/files/17459268/verifiableCertification1.json)
[verifiableCertification2.json](https://github.com/user-attachments/files/17459269/verifiableCertification2.json)
[verifiableCertification3.json](https://github.com/user-attachments/files/17459270/verifiableCertification3.json)
[verifiableCertification4.json](https://github.com/user-attachments/files/17459271/verifiableCertification4.json)
[verifiableCertification5.json](https://github.com/user-attachments/files/17459272/verifiableCertification5.json)
[verifiableCertification6.json](https://github.com/user-attachments/files/17459273/verifiableCertification6.json)
[verifiableCertification7.json](https://github.com/user-attachments/files/17459275/verifiableCertification7.json)
[verifiableCertification8.json](https://github.com/user-attachments/files/17459276/verifiableCertification8.json)

**Comment by hesusruiz** [comment]
I have some comments:

1. The Issuer of the Credential must be the signer of the Credential, to be alligned with the EUDI specifivations, especially when the serialization format is JWT or [SD-JWT-VC](https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-05.html#section-3.5).
2. There is no information about the source of each of the compliance items. For example, there is no information on the original CAB (Conformity Assessment Body) that issued the original ISO 27001 certification (in PDF format).
3. The information about the CAB and the actual Original Certification is important, because DEKRA does not verify that the service complies with ISO 27001, only the certification presented by the CSP is authentic and corresponds to the subject of the compliance item.

**Reply by noeguedek** [reply]
I agree with Jesus in 2 and 3. We should include the CAB that issues the original certificate and the identification of the original certificate if it has. I also suggest to remove the current id of compliance profile "Unique identifier for the compliance certification" since this id is locally to DOME and we couldn't reuse it for example for Gaia-X

**Reply by oriolcanades** [reply]
@noeguedek, could you please share a JSON example of what the result of these implementations could look like?

**Comment by wistefan** [comment]
I would also like to one question:
Why is the receiving company included in the subject, instead of binding the credential to the company as a holder?

**Reply by oriolcanades** [reply]
Could you share an example of what you expose?

**Comment by noeguedek** [comment]
I would like to add two more comments:
1. According to W3C VC data model we should use validFrom and validUntil tags (https://www.w3.org/TR/vc-data-model-2.0/#validity-periods) to indicate the validity period of a VC instead of issuanceDate and expirationDate
2. We are using productId of DOME database to unique identify the product in the VC but I think that can be a limitation if we want to use the same VC in Gaia-X.

**Reply by oriolcanades** [reply]
In response to point 1, I think following the same ontology is a good approach. I agree with your perspective: if we intend to use "validFrom", it would likely be better to change "expirationDate" to "validUntil", as "expirationDate" is not included in the definition at "https://www.w3.org/ns/credentials/v2".

**Comment by oriolcanades** [comment]
Another point we need to discuss is the possible use of the Issuer claim. In cases where the TSP delegates issuance to a third party, the TSP acts as an Attest Body but not as the Issuer. The Issuer should be the entity that signs the VC.

It may be necessary to include information about the TSP as an Attest Body within the CredentialSubject, such as:
- Attest Body (company)
- Company
- Product
- Compliance
